linker/Linker.Steps/MarkStep.cs

```diff
@@ -2053,6 +2086,246 @@ bool CheckReflectionMethod (Instruction instruction, string reflectionMethod)
+			if (previousPreviousInstruction == null)
+				return false;
+
+			if (previousInstruction.OpCode.Code != Code.Call || previousPreviousInstruction.OpCode.Code != Code.Ldtoken)
+				return false;
+
+			var creationType = (previousPreviousInstruction.Operand as TypeReference)?.Resolve ();
+
+			if (creationType == null)
+				return false;
+
+			// We can only detect the cases that use ctor arguments so we can assume we should only mark the default ctor
+			MarkDefaultConstructor (creationType);
+			return true;
+		}
+
+		/// <summary>
+		/// Note : Extra unused arguments are passed so that derived MarkStep's can try to do more sophisticated detection if they choose
+		/// </summary>
+		/// <param name="callingBody"></param>
+		/// <param name="method"></param>
+		/// <param name="callInstruction"></param>
+		/// <returns></returns>
+		protected virtual bool MarkCreationTypeOfCreateInstanceWithStringString (MethodBody callingBody, MethodReference method, Instruction callInstruction)
+		{
+			//
+			// Expected pattern is
+			// IL_0000: ldstr "test"
+			// IL_0005: ldstr "Mono.Linker.Tests.Cases.Reflection.Activator.StringOverload.SameAssembly+Foo"
+			// IL_000a: call class [mscorlib]System.Runtime.Remoting.ObjectHandle [mscorlib]System.Activator::CreateInstance(string, string)
+			//
+			var previousInstruction = callInstruction.Previous;
+			if (previousInstruction == null)
+				return false;
+			
+			var previousPreviousInstruction = previousInstruction.Previous;
+			if (previousPreviousInstruction == null)
+				return false;
+
+			if (previousInstruction.OpCode.Code != Code.Ldstr || previousPreviousInstruction.OpCode.Code != Code.Ldstr)
+				return false;
+
+			var typeName = previousInstruction.Operand as string;
+			var assemblyName = previousPreviousInstruction.Operand as string;
+
+			if (string.IsNullOrEmpty (typeName) || string.IsNullOrEmpty (assemblyName))
+				return false;
+
+			var assemblyDefinition = _context.GetAssemblies ().FirstOrDefault (asm => asm.Name.Name == assemblyName);
+			var createdType = assemblyDefinition?.MainModule.GetType (typeName.ToCecilName ());
+
+			if (createdType == null)
+				return false;
+
+			// We can only detect the cases that don't pass ctor arguments so it's safe to assume only default ctor is needed
+			MarkDefaultConstructor (createdType);
+			return true;
+		}
+
+		bool MarkCastTypeOfCreateInstanceWithType (MethodBody callingBody, MethodReference method, Instruction callInstruction)
+		{
+			var nextInstruction = callInstruction.Next;
+			if (nextInstruction == null)
+				return false;
+
+			if (nextInstruction.OpCode.Code != Code.Isinst && nextInstruction.OpCode.Code != Code.Castclass)
+				return false;
+			
+			var instanceBeingCastedToType = nextInstruction.Operand as TypeReference;
+			if (instanceBeingCastedToType == null)
+				return false;
+
+			var castType = BCL.Activator.ResolveCastType (instanceBeingCastedToType);
+
+			if (castType == null)
```

Comment:
Please add a test castType is a class

Reply:
Done.  Now looks like

```
			if (castType == null || castType.IsValueType)
				return false;
```

Reply:
Use `IsClass` as that's the only case the code can handle

Reply:
Switching to `IsClass` leads to the interface tests failing, which makes sense because interfaces are not classes and the code can handle interfaces.

Reply:
Didn't notice that so let's add || IsInterface